### PR TITLE
Fix status serialization / inappropriate sound effects

### PR DESCRIPTION
**Pull request by baughj**
## Description
This PR fixes statuses not persisting correctly after logoff (users logging back in would get a serialization error and never be able to login again), and also fixes a bug where a sound was being played on every tick.

## Related PRs
Should be merged after #142 .

## Checklist
- [X] Tested and working locally
- [ ] Reviewed
- [ ] Tested and working on dev server
- [ ] Deployed to staging

## How to QA
Cast Armachd.
Verify that there are no sounds playing over and over.
Logoff.
Log back in.
Note that the status is reapplied with the correct remaining duration.

## Impacted Areas in Hybrasyl
Statuses, user serialization.

### CODE REVIEW

File: HybrasylXML/XSD/Castables.xsd [hunk 1]
```diff
@@ -213,7 +213,7 @@
 
   <xs:complexType name="CastCostList">
     <xs:sequence>
-      <xs:element name="CastCost" type="actions:CastCost"/>
+      <xs:element name="CastCost" type="actions:CastCost"  minOccurs="1" maxOccurs="unbounded"/>
```
Comment (norrismiv) [inline HybrasylXML/XSD/Castables.xsd]: CastCost should not have a minimum occurrance, as there as spells with no casting cost.

File: HybrasylXML/XSD/Castables.xsd [hunk 2]
```diff
@@ -413,9 +413,9 @@
   <xs:complexType name="AddStatus">
     <xs:simpleContent>
       <xs:extension base="hyb:String8">
-        <xs:attribute name="Duration" use="optional" default="0" type="xs:int"/>
-        <xs:attribute name="Intensity" use="optional" default="1.0" type="xs:float"/>
-        <xs:attribute name="Speed" default="1.0" use="optional" type="xs:float"/>
+        <xs:attribute name="Duration"  default="0" type="xs:int"/>
+        <xs:attribute name="Intensity" default="1.0" type="xs:float"/>
+        <xs:attribute name="Speed" default="1.0" type="xs:float"/>
```
Comment (norrismiv) [inline HybrasylXML/XSD/Castables.xsd]: These should be optional, and not required.

File: HybrasylXML/XSD/Castables.xsd [hunk 3]
```diff
@@ -440,7 +440,6 @@
       <xs:element name="Restrictions" type="actions:Restrictions" minOccurs="0" maxOccurs="1" />
       <xs:element name="Effects" type="actions:Effects" minOccurs="0" maxOccurs="1" />
       <xs:element name="Script" type="xs:string" minOccurs="0" maxOccurs="1"/>
-      <xs:element name="Statuses" type="actions:Statuses" minOccurs="0" maxOccurs="1"/>
```
Comment (norrismiv) [inline HybrasylXML/XSD/Castables.xsd]: A castable can have multiple statuses. Not sure why this was removed.

File: HybrasylXML/XSD/hybrasylExtensions.cs [hunk 4]
```diff
@@ -222,20 +222,19 @@ public byte GetMaxLevelByClass(Class castableClass)
...
-            catch (NullReferenceException)
-            {
+
+            if (Effects?.Animations?.OnCast?.Player == null) return false;
+
+            var m = Effects.Animations.OnCast.Player.Where(e => e.Class.Contains(castClass));
+            if (m.Count() == 0)
+                m = Effects.Animations.OnCast.Player.Where(e => e.Class.Count == 0);
+
+            if (m.Count() == 0)
```
Comment (norrismiv) [inline HybrasylXML/XSD/hybrasylExtensions.cs]: This should remain singleordefault, as Where will return an enumerable, and single will return the actual object. SingleOrDefault will return properly. Default should be defined on the animation.

File: HybrasylXML/XSD/hybrasylExtensions.cs [hunk 5]
```diff
@@ -222,20 +222,19 @@ public byte GetMaxLevelByClass(Class castableClass)
...
                 return false;
-            }           
-            if (motion != null) return true;
-            return false;
+            else
+                motion = m.First();
```
Comment (norrismiv) [inline HybrasylXML/XSD/hybrasylExtensions.cs]: this is a result of moving away from singleordefault, and should be corrected.

File: HybrasylXML/XSD/hybrasylTypes.xsd [hunk 6]
```diff
@@ -190,12 +190,16 @@
...
-      <xs:enumeration value="Hostile"/> <!-- Hostile to players and can be attacked -->
-      <xs:enumeration value="Passive"/> <!-- Passive to players and cannot be attacked -->
-      <xs:enumeration value="Neutral"/> <!-- Passive to players and can be attacked -->
+      <xs:enumeration value="Hostile"/>
+      <!-- Hostile to players and can be attacked -->
+      <xs:enumeration value="Passive"/>
+      <!-- Passive to players and cannot be attacked -->
+      <xs:enumeration value="Neutral"/>
+      <!-- Passive to players and can be attacked -->
+
```
Comment (norrismiv) [inline HybrasylXML/XSD/hybrasylTypes.xsd]: please keep the commenting inline.

File: hybrasyl/CreatureStatus.cs [hunk 7]
```diff
@@ -71,18 +73,22 @@ public bool Check(ConditionInfo condition)
             return false;
         }
     }
-
+    
     public interface ICreatureStatus
     {
 
         string Name { get; }
+        string CastableName { get; }
```
Comment (norrismiv) [inline hybrasyl/CreatureStatus.cs]: A status should be independent of the castable. The castable should reference the status. Not sure why this is here.

File: hybrasyl/CreatureStatus.cs [hunk 8]
```diff
@@ -94,6 +100,13 @@ public interface ICreatureStatus
         void OnEnd();
 
     }
+   
+    public class StatusInfo
+    {
+        public string Name;
+        public string CastableName;
+        public double Remaining;
+    }
```
Comment (norrismiv) [inline hybrasyl/CreatureStatus.cs]: Is this a simple data structure? If so, Castable name should be removed, and proper getter/setter should be specified on the properties.

File: hybrasyl/CreatureStatus.cs [hunk 9]
```diff
@@ -105,11 +118,13 @@ public class CreatureStatus : ICreatureStatus
         public string UseCastRestrictions => XMLStatus.CastRestriction?.Use ?? string.Empty;
         public string ReceiveCastRestrictions => XMLStatus.CastRestriction?.Receive ?? string.Empty;
 
+        public StatusInfo Info => new StatusInfo() { Name = Name, CastableName = CastableName, Remaining = Remaining };
```
Comment (norrismiv) [inline hybrasyl/CreatureStatus.cs]: CastableName should not be applicable to a status.

File: hybrasyl/CreatureStatus.cs [hunk 10]
```diff
@@ -136,39 +151,38 @@ public class CreatureStatus : ICreatureStatus
...
-            int durationOverride = -1, int tickOverride = -1)
-        {
-            _init(xmlstatus, target, castable);
             Start = DateTime.Now;
             _durationOverride = durationOverride;
             _tickOverride = tickOverride;
         }
 
-        public CreatureStatus(Status xmlstatus, Creature target, Castable castable = null)
+        public CreatureStatus(Status xmlstatus, Creature target, Creature source=null, Castable castable = null)
```
Comment (norrismiv) [inline hybrasyl/CreatureStatus.cs]: It should not matter who applied the status.

File: hybrasyl/CreatureStatus.cs [hunk 11]
```diff
@@ -136,39 +151,38 @@ public class CreatureStatus : ICreatureStatus
...
         {
-            _init(xmlstatus, target, castable);
+            Target = target;
+            Source = source;
+            XMLStatus = xmlstatus;
```
Comment (norrismiv) [inline hybrasyl/CreatureStatus.cs]: What is XMLStatus? Proper property name would be XmlStatus. Even then it should just be 'Status'.

File: hybrasyl/CreatureStatus.cs [hunk 12]
```diff
@@ -136,39 +151,38 @@ public class CreatureStatus : ICreatureStatus
...
-                _durationOverride = addObj.Duration * xmlstatus.Duration;
+                _durationOverride = addObj.Duration != 0 ? addObj.Duration : xmlstatus.Duration;
                 _tickOverride = (int) Math.Floor(addObj.Speed * xmlstatus.Tick);
             }
         }
 
         private void ProcessSfx(ModifierEffect effect)
         {
-            if (effect.Sound?.Id != null)
+            if (effect.Sound?.Id != 0)
```
Comment (norrismiv) [inline hybrasyl/CreatureStatus.cs]: Sound is nullable. A null check on Id should apply.

File: hybrasyl/Enums.cs [hunk 13]
```diff
@@ -353,11 +353,11 @@ public enum ItemDropType
...
-            Alive = 0x01,
-            InExchange = 0x20,
-            InDialog = 0x40,
-            Casting = 0x100,
-            Pvp = 0x200,
+            Alive = 1,
+            InExchange = 2,
+            InDialog = 4,
+            Casting = 8,
+            Pvp = 16,
```
Comment (norrismiv) [inline hybrasyl/Enums.cs]: Why is this enum being reconfigured?

File: hybrasyl/Login.cs [hunk 14]
```diff
@@ -129,7 +130,7 @@ private void PacketHandler_0x03_Login(Client client, ClientPacket packet)
                     client.ConnectionId, name);
                 client.Redirect(redirect);
                 loginUser.Login.LastLogin = DateTime.Now;
-                loginUser.Login.LastLoginFrom = ((IPEndPoint) client.Socket.RemoteEndPoint).Address.ToString();
+                loginUser.Login.LastLoginFrom = ((IPEndPoint) client.Socket.RemoteEndPoint).Address.ToString();
```
Comment (norrismiv) [inline hybrasyl/Login.cs]: Please do not add unnecessary whitespace.

File: hybrasyl/Messaging/Statuses.cs [hunk 15]
```diff
@@ -21,7 +21,7 @@ public new static ChatCommandResult Run(User user, params string[] args)
         {
             if (Game.World.WorldData.TryGetValueByIndex(args[0], out Status status))
             {
-                user.ApplyStatus(new CreatureStatus(status, user, null));
+                user.ApplyStatus(new CreatureStatus(status, user, null, null));
```
Comment (norrismiv) [inline hybrasyl/Messaging/Statuses.cs]: This should be corrected, as "source" does not need to exist.

File: hybrasyl/NumberCruncher.cs [hunk 16]
```diff
@@ -145,7 +145,7 @@ public static double CalculateHeal(Castable castable, ModifierEffect effect, Cre
 
             if (effect?.Heal == null) return heal;
 
-            var statusAdd = castable?.Statuses?.Add?.Where(e => e.Value == statusName)?.ToList();
+            var statusAdd = castable?.Effects?.Statuses?.Add?.Where(e => e.Value == statusName)?.ToList();
             var intensity = statusAdd != null ? statusAdd[0].Intensity : 1;
```
Comment (norrismiv) [inline hybrasyl/NumberCruncher.cs]: this appears to only check the first intensity of a list of statuses. This must be corrected.

File: hybrasyl/Objects/Creature.cs [hunk 17]
```diff
@@ -547,7 +549,7 @@ public virtual bool UseCastable(Castable castObject, Creature target = null)
                     if (World.WorldData.TryGetValueByIndex<Status>(status.Value, out applyStatus))
                     {
                         ActivityLogger.Info($"UseCastable: {Name} casting {castObject.Name} - applying status {status.Value}");
-                        ApplyStatus(new CreatureStatus(applyStatus, tar, castObject));
+                        ApplyStatus(new CreatureStatus(applyStatus, tar, this, castObject));
```
Comment (norrismiv) [inline hybrasyl/Objects/Creature.cs]: "source" should not exist.

File: hybrasyl/Objects/Monster.cs [hunk 18]
```diff
@@ -261,6 +261,7 @@ public void Cast(Creature target)
             var castable = World.WorldData.Get<Castable>(creatureCastable.Value);
             if (target is Merchant) return;
             UseCastable(castable, target);
+            Condition.Casting = false;
```
Comment (norrismiv) [inline hybrasyl/Objects/Monster.cs]: This should be handled in the UseCastable method.

File: hybrasyl/Objects/User.cs [hunk 19]
```diff
@@ -1155,12 +1167,20 @@ internal void UseSpell(byte slot, uint target = 0)
         /// <returns>True or false depending on success.</returns>
         public bool ProcessCastingCost(Castable castable)
         {
-            if (castable.CastCosts?.CastCost == null) return true;
+            if (castable.CastCosts.Count == 0) return true;
+
+            var costStruct = castable.CastCosts.Where(e => e.Class.Contains((Class)Class));
```
Comment (norrismiv) [inline hybrasyl/Objects/User.cs]: this variable naming is confusing. This should be simply `var costs` as this is an enumerable.

File: hybrasyl/Objects/User.cs [hunk 20]
```diff
@@ -1155,12 +1167,20 @@ internal void UseSpell(byte slot, uint target = 0)
...
+
+            if (costStruct.Count() == 0)
+                return true;
 
-            var costs = castable.CastCosts.CastCost;
             uint reduceHp = 0;
             uint reduceMp = 0;
-            int removeNumItems = 0;
+            bool hasItemCost = true;
+            var costs = costStruct.First();
```
Comment (norrismiv) [inline hybrasyl/Objects/User.cs]: this will only apply the first cast cost, and is handled improperly.

File: hybrasyl/Objects/User.cs [hunk 21]
```diff
@@ -2086,7 +2101,7 @@ public override void RegenerateMp(double mp, Creature regenerator = null)
                     Stats.Hp = 1;
                     var handler = Game.Config.Handlers?.Death?.Coma;
                     if (handler != null && World.WorldData.TryGetValueByIndex(handler.Value, out Status status))
-                        ApplyStatus(new CreatureStatus(status, this, null));
+                        ApplyStatus(new CreatureStatus(status, this, attacker, null));
```
Comment (norrismiv) [inline hybrasyl/Objects/User.cs]: "source" should not exist.

File: hybrasyl/Objects/User.cs [hunk 22]
```diff
@@ -2142,7 +2159,7 @@ public void AssailAttack(Direction direction, Creature target = null)
             Enqueue(assail.Packet());
             PlaySound(soundId);
             SendAnimation(assail.Packet());
-            PlaySound(soundId);
+            PlaySound(soundId);
```
Comment (norrismiv) [inline hybrasyl/Objects/User.cs]: please do not add unnecessary whitespace

File: HybrasylXML/XSD/Castables.xsd (hunk 2 above)
Comment (baughj) [reply HybrasylXML/XSD/Castables.xsd]: Optional with a default value has the same effect.

File: HybrasylXML/XSD/Castables.xsd (hunk 3 above)
Comment (baughj) [reply HybrasylXML/XSD/Castables.xsd]: Right now it is duplicated. This is removing this one in favor of the block under `Effects`.

File: hybrasyl/CreatureStatus.cs (hunk 10 above)
Comment (baughj) [reply hybrasyl/CreatureStatus.cs]: Formulas can be used in status damage. There are several examples of formulas that use `$CASTER` values. This ensures they work.

File: hybrasyl/CreatureStatus.cs (hunk 12 above)
Comment (baughj) [reply hybrasyl/CreatureStatus.cs]: Sound is not nullable as exists currently. This check was failing, which is why it was changed. `Sound` always has a default value.

File: hybrasyl/Enums.cs (hunk 13 above)
Comment (baughj) [reply hybrasyl/Enums.cs]: For clarity and to make sure it aligns cleanly.

File: hybrasyl/NumberCruncher.cs (hunk 16 above)
Comment (baughj) [reply hybrasyl/NumberCruncher.cs]: You cannot add the same status multiple times. Damage from each tick would be checked against its own intensity, since it is possible to add multiple statuses with different intensities.

File: hybrasyl/Messaging/Statuses.cs (hunk 15 above)
Comment (baughj) [reply hybrasyl/Messaging/Statuses.cs]: See above.

File: hybrasyl/Objects/Creature.cs (hunk 17 above)
Comment (baughj) [reply hybrasyl/Objects/Creature.cs]: See above.

File: hybrasyl/Objects/User.cs (hunk 20 above)
Comment (baughj) [reply hybrasyl/Objects/User.cs]: The cast cost enumerated is either the first one that specifies a class, or a default (no class). Why would you need to process multiple cast cost structures for the same class? There is no need for that duplication.

File: hybrasyl/Objects/User.cs (hunk 21 above)
Comment (baughj) [reply hybrasyl/Objects/User.cs]: See above.

File: HybrasylXML/XSD/Castables.xsd (hunk 1 above)
Comment (baughj) [reply HybrasylXML/XSD/Castables.xsd]: Will fix, unintentional error.

File: hybrasyl/CreatureStatus.cs (hunk 7 above)
Comment (baughj) [reply hybrasyl/CreatureStatus.cs]: Included due to current accessibility in formulas. Also, the castable itself is needed for the `Intensity` value, though that could also be serialized to remove that specific dependency.

File: hybrasyl/CreatureStatus.cs (hunk 8 above)
Comment (baughj) [reply hybrasyl/CreatureStatus.cs]: Will fix.

File: hybrasyl/CreatureStatus.cs (hunk 9 above)
Comment (baughj) [reply hybrasyl/CreatureStatus.cs]: See above.

File: hybrasyl/CreatureStatus.cs (hunk 11 above)
Comment (baughj) [reply hybrasyl/CreatureStatus.cs]: This would have been a comment to include in the original review for statuses (note that this is used extensively here and did not appear in this commit).

However, since it does not follow convention, I have fixed it.

`XmlStatus` is a `Statuses.Status` structure.

File: hybrasyl/Objects/Monster.cs (hunk 18 above)
Comment (baughj) [reply hybrasyl/Objects/Monster.cs]: Good point. Fixed.

File: HybrasylXML/XSD/Castables.xsd (hunk 2 above)
Comment (norrismiv) [reply HybrasylXML/XSD/Castables.xsd]: Yes, except optional was removed.

File: HybrasylXML/XSD/Castables.xsd (hunk 3 above)
Comment (norrismiv) [reply HybrasylXML/XSD/Castables.xsd]: Fair enough.

File: hybrasyl/Objects/User.cs (hunk 22 above)
Comment (baughj) [reply hybrasyl/Objects/User.cs]: Not sure why this whitespace was added. Will be more careful in the future.

File: hybrasyl/Login.cs (hunk 14 above)
Comment (baughj) [reply hybrasyl/Login.cs]: Same as below, not sure why this whitespace was added. Fixed.

File: hybrasyl/CreatureStatus.cs (hunk 7 above)
Comment (norrismiv) [reply hybrasyl/CreatureStatus.cs]: Fair enough, though the point of statuses being separate was to remove the dependency on castables.

File: hybrasyl/CreatureStatus.cs (hunk 10 above)
Comment (norrismiv) [reply hybrasyl/CreatureStatus.cs]: I'd like an example of a status that uses a caster based formula. From my understanding is that status formulas should not have any reliance on the specific caster.

File: hybrasyl/CreatureStatus.cs (hunk 12 above)
Comment (norrismiv) [reply hybrasyl/CreatureStatus.cs]: Looking at that block, sound is indeed nullable.

File: hybrasyl/Enums.cs (hunk 13 above)
Comment (norrismiv) [reply hybrasyl/Enums.cs]: Fair enough

File: HybrasylXML/XSD/Castables.xsd (hunk 2 above)
Comment (baughj) [reply HybrasylXML/XSD/Castables.xsd]: Right, because in my mind if a default results in a value being set, `optional` is unclear / unneeded. Since it has no effect on the generated code (optional for instance doesn't mean it is a `float?`) it seems spurious to me. I can add it back though if you feel it is clearer.

File: hybrasyl/Messaging/Statuses.cs (hunk 15 above)
Comment (norrismiv) [reply hybrasyl/Messaging/Statuses.cs]: See above as well.

File: hybrasyl/NumberCruncher.cs (hunk 16 above)
Comment (norrismiv) [reply hybrasyl/NumberCruncher.cs]: Right, but you can add multiple separate statuses from a castable. So this would only check the first intensity, and ignore the rest of the Add statuses.

File: hybrasyl/Objects/Creature.cs (hunk 17 above)
Comment (norrismiv) [reply hybrasyl/Objects/Creature.cs]: See above as well.

File: hybrasyl/CreatureStatus.cs (hunk 7 above)
Comment (baughj) [reply hybrasyl/CreatureStatus.cs]: Sure, this can be `String.Empty` though (note that the CreatureStatus allows this). Really, it is only needed if used in the formula. It is included here to handle the more generic case of a castable having a defined intensity on a reused status.

File: hybrasyl/Objects/User.cs (hunk 20 above)
Comment (norrismiv) [reply hybrasyl/Objects/User.cs]: Fair enough. This should be moved to the above comment. as .First() at the end for clarity.

File: hybrasyl/Objects/User.cs (hunk 21 above)
Comment (norrismiv) [reply hybrasyl/Objects/User.cs]: See above as well.

File: hybrasyl/CreatureStatus.cs (hunk 10 above)
Comment (baughj) [reply hybrasyl/CreatureStatus.cs]: Statuses are not really implemented 100% yet on the data side, so there is not one currently. There are many examples of valid formulas that would need this information. Are you saying you'd rather have the potential for multiple `NullReferenceExceptions` in damage calculation rather than including an object reference in a status? That seems very strange to me. 

There is no reliance - currently - but we decided to allow formulas in statuses, so, this makes that work correctly with intended results and no surprises. If we only want statuses to have a specific set of formulas or to have no formulas, that is a different question altogether, but it seems like bad behavior to say "You can only use a subset of the standard formula variables here" to XML creators.

File: hybrasyl/CreatureStatus.cs (hunk 12 above)
Comment (baughj) [reply hybrasyl/CreatureStatus.cs]: It is not a nullable type in the structure `xsd2code` generates, is what I mean. My expectation that it is by use of `?` is wrong. `Status.Designer.cs` line 1210; also note it is always initialized in `ModifierEffect` (line 1116).

If you have a better suggestion for how to handle this case we see elsewhere, I welcome it. We may need to make more fields nullable as it is untenable for many things like this to have a default of `0` as `0` is in fact a valid sound. This fix in particular though is a workaround to the `weird sound playing every status tick` bug because previously this is never null and will always result in a sound playing. Cast `Armachd` and observe the results with the null check in place.

File: hybrasyl/CreatureStatus.cs (hunk 10 above)
Comment (norrismiv) [reply hybrasyl/CreatureStatus.cs]: Reliance on caster variables in status formulas will introduce a slew of complexity that should be accounted for here. If this is the path that you'd like to take, then so be it, but I'm almost guaranteeing that it becomes very messy.

For example, how to you maintain the caster data at the time of spell cast versus just a solid reference to the caster? How does this work when logging off and logging back on? What happens to the status during a leveling cycle and the caster levels up? Or changes equipment? Or changes stats? Or while master buys stats?

This is just a sampling of the variables that need to be accounted for if this is the path you wish to take with this.

File: hybrasyl/NumberCruncher.cs (hunk 16 above)
Comment (baughj) [reply hybrasyl/NumberCruncher.cs]: Please note I am finding the intensity for the current effect (`e.Value == statusName`). I do not need to care about other intensities here; they are handled on their own. `CalculateHeal / Damage` is run individually for each tick of a status. For instance, if I had a heal over time applied, I would rely on its defined intensity in the castable. I don't need to know the intensity of anything else (say, a poison).

There should never be an instance of adding the same named status twice with differing intensities; that makes no sense (and also isn't possible with current client limitations)

File: hybrasyl/CreatureStatus.cs (hunk 10 above)
Comment (baughj) [reply hybrasyl/CreatureStatus.cs]: Fair point about logging on and off. What about evaluating the formula once on application and storing the result for future use? Otherwise we're basically saying that statuses can only have simple damage, which seems pretty limiting.

**Comment by moseleyc** [comment]
Status timer doesn't appear to be working after logging out, and logging back in. Even a recast of the same spell appears to have no effect. My test is with a 10 second Armachd, and after logging out /in, it's still there after at least a couple minutes.

**Comment by baughj** [comment]
@moseleyc fixed, and verified fixed.
@norrismiv status serialization handled as discussed.

**Comment by moseleyc** [comment]
Getting NullReferenceException:

2018-02-19 19:33:50,914 [6] ERROR Hybrasyl.Client EXCEPTION IN HANDLING: 0x03: Newtonsoft.Json.JsonSerializationException: Error setting value to 'Statuses' on 'Hybrasyl.Objects.User'. ---> System.NullReferenceException: Object reference not set to an instance of an object.
   at Hybrasyl.Objects.User.SendSystemMessage(String p) in C:\Users\Agrus\Desktop\Darkages-Workspace\ServerSplitNonGit\hybrasylserver-moar-status-fixes\hybrasyl\Objects\User.cs:line 3709
   at Hybrasyl.CreatureStatus.ProcessSfx(ModifierEffect effect) in C:\Users\Agrus\Desktop\Darkages-Workspace\ServerSplitNonGit\hybrasylserver-moar-status-fixes\hybrasyl\CreatureStatus.cs:line 241
   at Hybrasyl.CreatureStatus.ProcessFullEffects(ModifierEffect effect, Boolean RemoveStatBonuses) in C:\Users\Agrus\Desktop\Darkages-Workspace\ServerSplitNonGit\hybrasylserver-moar-status-fixes\hybrasyl\CreatureStatus.cs:line 340
   at Hybrasyl.CreatureStatus._processStart() in C:\Users\Agrus\Desktop\Darkages-Workspace\ServerSplitNonGit\hybrasylserver-moar-status-fixes\hybrasyl\CreatureStatus.cs:line 348
   at Hybrasyl.CreatureStatus.OnStart() in C:\Users\Agrus\Desktop\Darkages-Workspace\ServerSplitNonGit\hybrasylserver-moar-status-fixes\hybrasyl\CreatureStatus.cs:line 162
   at Hybrasyl.Objects.Creature.ApplyStatus(ICreatureStatus status, Boolean sendUpdates) in C:\Users\Agrus\Desktop\Darkages-Workspace\ServerSplitNonGit\hybrasylserver-moar-status-fixes\hybrasyl\Objects\Creature.cs:line 419
   at Hybrasyl.Objects.Creature.set_Statuses(List`1 value) in C:\Users\Agrus\Desktop\Darkages-Workspace\ServerSplitNonGit\hybrasylserver-moar-status-fixes\hybrasyl\Objects\Creature.cs:line 68
   at SetStatuses(Object , Object )
   at Newtonsoft.Json.Serialization.DynamicValueProvider.SetValue(Object target, Object value)
   --- End of inner exception stack trace ---
   at Newtonsoft.Json.Serialization.DynamicValueProvider.SetValue(Object target, Object value)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.SetPropertyValue(JsonProperty property, JsonConverter propertyConverter, JsonContainerContract containerContract, JsonProperty containerProperty, JsonReader reader, Object target)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.PopulateObject(Object newObject, JsonReader reader, JsonObjectContract contract, JsonProperty member, String id)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateObject(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, Object existingValue)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.Deserialize(JsonReader reader, Type objectType, Boolean checkAdditionalContent)
   at Newtonsoft.Json.JsonSerializer.DeserializeInternal(JsonReader reader, Type objectType)
   at Newtonsoft.Json.JsonConvert.DeserializeObject(String value, Type type, JsonSerializerSettings settings)
   at Newtonsoft.Json.JsonConvert.DeserializeObject[T](String value, JsonSerializerSettings settings)
   at Hybrasyl.World.TryGetUser(String name, User& userobj) in C:\Users\Agrus\Desktop\Darkages-Workspace\ServerSplitNonGit\hybrasylserver-moar-status-fixes\hybrasyl\World.cs:line 208
   at Hybrasyl.Login.PacketHandler_0x03_Login(Client client, ClientPacket packet) in C:\Users\Agrus\Desktop\Darkages-Workspace\ServerSplitNonGit\hybrasylserver-moar-status-fixes\hybrasyl\Login.cs:line 105
   at Hybrasyl.Client.FlushReceiveBuffer() in C:\Users\Agrus\Desktop\Darkages-Workspace\ServerSplitNonGit\hybrasylserver-moar-status-fixes\hybrasyl\Client.cs:line 512

**Comment by baughj** [comment]
Fixed.

**Comment by moseleyc** [comment]
Status still not being removed at end of time. Runs into Negative digits.

[hybrasylPartialLogEntries.txt](https://github.com/hybrasyl/server/files/1738918/hybrasylPartialLogEntries.txt)

**Comment by baughj** [comment]
Confirmed fixed for all scenarios.

**Comment by moseleyc** [comment]
Confirmed, issues fixed.
